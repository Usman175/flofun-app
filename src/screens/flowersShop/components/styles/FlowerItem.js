import EStyleSheet from 'react-native-extended-stylesheet'
import { StyleSheet } from 'react-native'

const styles = EStyleSheet.create({
  root: {
    width: '95%',
    height: 200,
    marginTop: '1.5%',
    backgroundColor: '$whiteColor',
    flexDirection: 'column',
  },
  lastFlower: {
    marginBottom: '1.5%',
  },
  imageSection: {
    flex: 1,
  },
  image: {
    flex: 1,
  },
  flowerDesc: {
    flexDirection: 'row',
    height: 45,
    alignItems: 'center',
    backgroundColor: '$quantityColor',
  },
  titleSection: {
    height: 45,
    width: '55%',
    justifyContent: 'center',
    backgroundColor: '$grayColor',
  },
  priceSection: {
    height: 45,
    width: '30%',
    justifyContent: 'center',
    alignItems: 'center',
    backgroundColor: '$purpleColor',
  },
  quantitySection: {
    height: 45,
    backgroundColor: '$quantityColor',
    flexDirection: 'row',
    alignItems: 'center',
    justifyContent: 'center',
    width: '0%',
  },
  quantityText: {
    fontSize: 20,
    fontFamily: 'montserratBold',
    color: '$purpleColor',
    paddingHorizontal: 15,
    backgroundColor: 'transparent',
  },
  quantitySelectorButton: {
    backgroundColor: '$purpleDarkColor',
    width: 35,
    height: 35,
    alignItems: 'center',
    justifyContent: 'center',
    borderRadius: 4,
  },
  quantitySelectorButtonText: {
    fontSize: 30,
    lineHeight: 30,
    color: '#f4efff',
  },
  addToCart: {
    height: 45,
    width: '15%',
    justifyContent: 'center',
    alignItems: 'center',
    backgroundColor: '$purpleDarkColor',
  },
  title: {
    fontSize: 15,
    fontFamily: 'montserratBold',
    color: '$whiteColor',
  },
  subTitle: {
    fontSize: 12,
    fontFamily: 'montserrat',
    color: '$whiteColor',
  },
  price: {
    fontSize: 16,
    fontFamily: 'montserratMedium',
    color: '$whiteColor',
  },
  priceSelected: {
    backgroundColor: '$greenColor',
  },
  modal: {
    width: 300,
    height: 400,
    backgroundColor: '$whiteColor',
    justifyContent: 'center',
    alignItems: 'center',
  },
  modalContainer: {
    ...StyleSheet.absoluteFillObject,
    backgroundColor: 'transparent',
    alignItems: 'center',
    justifyContent: 'center',
  },
  backgroundModal: {
    ...StyleSheet.absoluteFillObject,
    backgroundColor: 'rgba(0,0,0,0.5)',
  },
})

export default styles
