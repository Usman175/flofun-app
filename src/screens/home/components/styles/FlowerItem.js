import EStyleSheet from 'react-native-extended-stylesheet'

const styles = EStyleSheet.create({
  root: {
    width: '95%',
    height: 160,
    marginTop: '1.5%',
    backgroundColor: '$purpleColor',
    flexDirection: 'row',
  },
  imageSection: {
    flex: 0.7,
  },
  image: {
    flex: 1,
  },
  flowerDesc: {
    flex: 0.3,
    flexDirection: 'column',
  },
  titleSection: {
    flex: 0.7,
    paddingLeft: 5,
    paddingTop: 5,
    backgroundColor: '$grayColor',
  },
  priceSection: {
    flex: 0.3,
    justifyContent: 'center',
    alignItems: 'center',
    backgroundColor: '$purpleColor',
  },
  title: {
    fontSize: 15,
    fontFamily: 'montserratBold',
    color: '$whiteColor',
  },
  subTitle: {
    fontSize: 12,
    fontFamily: 'montserrat',
    color: '$whiteColor',
  },
  price: {
    fontSize: 16,
    fontFamily: 'montserratMedium',
    color: '$whiteColor',
  },
  priceSelected: {
    backgroundColor: '$greenColor',
  },
})

export default styles
