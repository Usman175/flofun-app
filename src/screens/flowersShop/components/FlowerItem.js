import React, { Component } from 'react'
import {
  View,
  Text,
  Image,
  TouchableOpacity,
  TouchableWithoutFeedback,
  Animated,
} from 'react-native'
import { MaterialIcons } from '@expo/vector-icons'

import styles from './styles/FlowerItem'

export default class FlowerItem extends Component {
  constructor(props) {
    super(props)

    this._animatedWidth = new Animated.Value(0)
    this._animatedQuantity = new Animated.Value(0)
    this.state = {
      isOpen: false,
      quantity: 1,
    }
  }

  closeQuantitySelector = () => {
    this.setState({ isOpen: false })
    Animated.spring(this._animatedWidth, {
      toValue: 0,
    }).start()
  }

  changeQuantity = change => {
    const quantity = this.state.quantity + change

    if (quantity === 0) {
      this.closeQuantitySelector()
    } else {
      Animated.sequence([
        Animated.timing(this._animatedQuantity, {
          toValue: -3,
          duration: 100,
        }),
        Animated.timing(this._animatedQuantity, {
          toValue: 3,
          duration: 100,
        }),
        Animated.timing(this._animatedQuantity, {
          toValue: 0,
          duration: 100,
        }),
      ]).start()
      this.setState({ quantity })
    }
  }

  handleSelectFlower = flower => {
    if (this.state.isOpen) {
      this.closeQuantitySelector()
      this.props.selectFlower({ ...flower, quantity: this.state.quantity })
    } else {
      this.setState({ isOpen: true })
      Animated.spring(this._animatedWidth, {
        toValue: 1,
      }).start()
    }
  }

  renderQuantitySelector = () => {
    const { flower, selected } = this.props
    const quantity = this.state.quantity

    const containerStyles = [
      styles.quantitySection,
      {
        width: this._animatedWidth.interpolate({
          inputRange: [0, 1],
          outputRange: ['0%', '55%'],
          extrapolate: 'clamp',
        }),
        transform: [{ scale: this._animatedWidth }],
      },
    ]

    const titleContainerStyles = [
      styles.titleSection,
      {
        width: this._animatedWidth.interpolate({
          inputRange: [0, 1],
          outputRange: ['55%', '0%'],
          extrapolate: 'clamp',
        }),
        paddingLeft: this._animatedWidth.interpolate({
          inputRange: [0, 1],
          outputRange: [5, 0],
          extrapolate: 'clamp',
        }),
      },
    ]

    const quantityStyles = [
      styles.quantityText,
      {
        transform: [
          {
            translateX: this._animatedQuantity,
          },
        ],
      },
    ]

    return (
      <View style={styles.flowerDesc}>
        <Animated.View style={titleContainerStyles}>
          <Text ellipsizeMode="tail" numberOfLines={2} style={styles.title}>{flower.title}</Text>
        </Animated.View>
        <View style={[styles.priceSection, selected ? styles.priceSelected : null]}>
          <Text style={styles.price}>$ {flower.price}</Text>
        </View>
        <Animated.View style={containerStyles}>
          <TouchableOpacity
            style={styles.quantitySelectorButton}
            onPress={() => this.changeQuantity(-1)}
          >
            <Text style={styles.quantitySelectorButtonText}>-</Text>
          </TouchableOpacity>
          <Animated.Text style={quantityStyles}>
            {quantity}
          </Animated.Text>
          <TouchableOpacity
            style={styles.quantitySelectorButton}
            onPress={() => this.changeQuantity(1)}
          >
            <Text style={styles.quantitySelectorButtonText}>+</Text>
          </TouchableOpacity>
        </Animated.View>
        <View
          style={[styles.addToCart, this.state.isOpen || selected ? styles.priceSelected : null]}
        >
          <TouchableOpacity onPress={() => this.handleSelectFlower(flower)}>
            {selected
              ? <MaterialIcons name="check-circle" size={30} color="#fff" />
              : <MaterialIcons name="add-shopping-cart" size={30} color="#fff" />}
          </TouchableOpacity>
        </View>
      </View>
    )
  }

  render() {
    const { flower, lastElement } = this.props

    return (
      <View style={[styles.root, lastElement && styles.lastFlower]}>
        <View style={styles.imageSection}>
          <TouchableWithoutFeedback onPress={this._openInImageGallery}>
            <Image source={{ uri: flower.image }} style={styles.image} />
          </TouchableWithoutFeedback>
        </View>
        {this.renderQuantitySelector()}
      </View>
    )
  }
}
