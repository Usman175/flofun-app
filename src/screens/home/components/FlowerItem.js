import React, { Component } from 'react'
import { View, Text, Image, TouchableOpacity } from 'react-native'
import styles from './styles/FlowerItem'

class FlowerItem extends Component {
  render() {
    const { flower, selectFlower, selected } = this.props

    return (
      <View style={styles.root}>
        <View style={styles.imageSection}>
          <Image
            source={{ uri: flower.image }}
            style={styles.image}
          />
        </View>
        <View style={styles.flowerDesc}>
          <View style={styles.titleSection}>
            <Text style={styles.title}>{flower.title}</Text>
            <Text style={styles.subTitle}>{flower.description}</Text>
          </View>
          <View style={[styles.priceSection, selected ? styles.priceSelected : null]}>
            <TouchableOpacity onPress={() => selectFlower(flower)}>
              <Text style={styles.price}>S./ {flower.price}</Text>
            </TouchableOpacity>
          </View>
        </View>
      </View>
    )
  }
}

export default FlowerItem
