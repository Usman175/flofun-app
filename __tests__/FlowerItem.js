import 'react-native'
import React from 'react'
import { shallow } from 'enzyme'
import toJson from 'enzyme-to-json'
import EStyleSheet from 'react-native-extended-stylesheet'
import Colors from '../constants/colors'
import FlowerItem from '../src/screens/home/components/FlowerItem'

beforeAll(() => {
  EStyleSheet.build(Colors)
})

describe('FlowerItem', () => {
  test('renders correctly', () => {
    const flower = shallow(
      <FlowerItem
        flower={{
          title: 'foo flower',
          image: 'https://cdn.arstechnica.net/wp-content/uploads/2016/02/5718897981_10faa45ac3_b-640x624.jpg',
          price: 32.00,
        }}
      />
    )

    expect(toJson(flower)).toMatchSnapshot()
  })
})
