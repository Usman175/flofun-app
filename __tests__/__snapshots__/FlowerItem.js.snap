// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FlowerItem renders correctly 1`] = `
<View
  style={
    Array [
      Object {
        "backgroundColor": "#574E6D",
        "flexDirection": "row",
        "height": 160,
        "marginTop": 13.34,
        "width": 712.5,
      },
      undefined,
    ]
  }
>
  <View
    style={
      Object {
        "flex": 0.7,
      }
    }
  >
    <Image
      source={
        Object {
          "uri": "https://cdn.arstechnica.net/wp-content/uploads/2016/02/5718897981_10faa45ac3_b-640x624.jpg",
        }
      }
      style={
        Object {
          "flex": 1,
        }
      }
    />
  </View>
  <View
    style={
      Object {
        "flex": 0.3,
        "flexDirection": "column",
      }
    }
  >
    <View
      style={
        Object {
          "backgroundColor": "#4B586E",
          "flex": 0.7,
          "paddingLeft": 5,
          "paddingTop": 5,
        }
      }
    >
      <Text
        accessible={true}
        allowFontScaling={true}
        ellipsizeMode="tail"
        style={
          Object {
            "color": "#fff",
            "fontFamily": "montserratBold",
            "fontSize": 15,
          }
        }
      >
        foo flower
      </Text>
      <Text
        accessible={true}
        allowFontScaling={true}
        ellipsizeMode="tail"
        style={
          Object {
            "color": "#fff",
            "fontFamily": "montserrat",
            "fontSize": 12,
          }
        }
      />
    </View>
    <View
      style={
        Array [
          Object {
            "alignItems": "center",
            "backgroundColor": "#574E6D",
            "flex": 0.3,
            "justifyContent": "center",
          },
          null,
        ]
      }
    >
      <TouchableOpacity
        activeOpacity={0.2}
        focusedOpacity={0.7}
        onPress={[Function]}
      >
        <Text
          accessible={true}
          allowFontScaling={true}
          ellipsizeMode="tail"
          style={
            Object {
              "color": "#fff",
              "fontFamily": "montserratMedium",
              "fontSize": 16,
            }
          }
        >
          S./ 
          32
        </Text>
      </TouchableOpacity>
    </View>
  </View>
</View>
`;
